### PR TITLE
Add cleanup function

### PR DESCRIPTION
**Pull request by yuriilisovskyi**
Added a cleanup function that checks if there are any bridges, VRFs or Qdiscs left and removes them. For now, it should be used at the end of the test and it is planned to convert it to pytest fixture in the future (in order to call it automatically after the test).

### CODE REVIEW

File: DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/tc/linux/linux_tc_qdisc_impl.py [hunk 1]
```diff
@@ -38,6 +38,8 @@ def format_show(self, command, *argv, **kwarg):
         """
         params = kwarg["params"]
         cmd = "tc qdisc {} ".format(command)
+        if "options" in params:
+            cmd = "tc " + params["options"] + " qdisc {} ".format(command)
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbedLib/src/dent_os_testbed/lib/tc/linux/linux_tc_qdisc_impl.py]: Since you are updating this class it would be nice to add the `parse_show` function here (see `LinuxIpLinkImpl` class for reference):
```python
def parse_show(self, command, output, *argv, **kwarg):
    return json.loads(output)
```
This will allow you to use the `parse_output` argument when calling `TcQdisc.show()`:
```python
out = await TcQdisc.show(input_data=[{dev.host_name: [
    {"options": "-j"}
]}], parse_output=True)
assert out[0][dev.host_name]["rc"] == 0
print(out[0][dev.host_name]["parsed_output"])  # { ... }
```

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 2]
```diff
@@ -73,6 +76,30 @@ async def tb_clean_config(testbed):
     check_asyncio_results(results, "tb_clean_config")
 
 
+async def tb_post_test_cleanup(testbed):
+    logger = AppLogger(DEFAULT_LOGGER)
+    dev = await tb_get_all_devices(testbed)
+    dev = dev[0]
+    logger.info("Clearing TC")
+    out = await TcQdisc.show(input_data=[{dev.host_name: [{"options": "-j"}]}])
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: Please use `parse_output=True` instead of parsing json yourself. See my other comment.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 3]
```diff
@@ -73,6 +76,30 @@ async def tb_clean_config(testbed):
...
+    qdiscs_info = json.loads(out[0][dev.host_name]["result"])
+    for qdisc_obj in qdiscs_info:
+        if qdisc_obj["kind"] != "noqueue":
+            await TcQdisc.delete(input_data=[{dev.host_name: [{"dev": qdisc_obj["dev"], "direction": qdisc_obj["kind"]}]}])
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: You can use a list comprehension here and only wait on one `TcQdisc.delete` call:
```python
await TcQdisc.delete(input_data=[{dev.host_name: [
    {"dev": qdisc_obj["dev"], "direction": qdisc_obj["kind"]}
    for qdisc_obj in qdiscs_info if qdisc_obj["kind"] != "noqueue"
]}])
```

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 4]
```diff
@@ -73,6 +76,30 @@ async def tb_clean_config(testbed):
     check_asyncio_results(results, "tb_clean_config")
 
 
+async def tb_post_test_cleanup(testbed):
+    logger = AppLogger(DEFAULT_LOGGER)
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: Please add a docstring as well

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 5]
```diff
@@ -5,7 +5,9 @@
 import re
 import pytest
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: Are you going to add a conftest.py with the new fixture in the scope of this PR?

**Review by anton7811 — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 6]
```diff
@@ -73,6 +76,47 @@ async def tb_clean_config(testbed):
...
+    """Removes all non-default qdiscs created during test. Can be used separately or by using `cleanup_qdiscs` fixture."""
+    logger = AppLogger(DEFAULT_LOGGER)
+    dev = await tb_get_all_devices(testbed)
+    dev = dev[0]
+    logger.info("Clearing TC")
+    out = await TcQdisc.show(input_data=[{dev.host_name: [{"options": "-j"}]}], parse_output=True)
+    qdiscs_info = out[0][dev.host_name]["parsed_output"]
+    for qdisc_obj in qdiscs_info:
+        if qdisc_obj["kind"] != "noqueue":
+            await TcQdisc.delete(input_data=[{dev.host_name: [{"dev": qdisc_obj["dev"], "direction": qdisc_obj["kind"]}]}])
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: Isn`t this line too long? The length should not exceed 120 characters.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 7]
```diff
@@ -73,6 +76,47 @@ async def tb_clean_config(testbed):
     check_asyncio_results(results, "tb_clean_config")
 
 
+async def tb_post_test_cleanup_qdiscs(testbed):
+    """Removes all non-default qdiscs created during test. Can be used separately or by using `cleanup_qdiscs` fixture."""
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: Same for this docstring - max length should not exceed 120 chars

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/conftest.py [hunk 8]
```diff
@@ -124,3 +130,21 @@ def pytest_runtest_makereport(item, call):
     docstring = getattr(test_fn, "__doc__")
     if docstring:
         print(docstring)
+
+
+@pytest.fixture
+def cleanup_qdiscs(testbed):
+    yield
+    asyncio.run(tb_post_test_cleanup_qdiscs(testbed))
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/conftest.py]: can we use `async` fixtures?

**Comment by SerhiyBoikoPLV** [comment]
besides some minor comments this looks good

**Comment by SerhiyBoikoPLV** [comment]
We also need a fixture for the Ixia sessions (https://github.com/dentproject/testing/issues/115). Can we add it in this PR?

**Comment by anton7811** [comment]
Please fix DCO check and do all changes based on the master branch.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/conftest.py (hunk 8 above)
Comment (anton7811) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/conftest.py]: we have to

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 9]
```diff
@@ -73,6 +76,47 @@ async def tb_clean_config(testbed):
...
+    logger.info("Clearing bridges")
+    out = await IpLink.show(input_data=[{dev.host_name: [{"link_type": "bridge", "cmd_options": "-j"}]}], parse_output=True)
+    bridges_info = out[0][dev.host_name]["parsed_output"]
+    if bridges_info:
+        await IpLink.delete(input_data=[{dev.host_name: [
+            {"device": bridge_obj["ifname"]} for bridge_obj in bridges_info]}
+        ])
+
+
+async def tb_post_test_cleanup_vrfs(testbed):
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: @SerhiyBoikoPLV  Do we need cleanup IPs as well?

**Review by anton7811 — CHANGES_REQUESTED** [review]

**Comment by yuriilisovskyi** [comment]
> We also need a fixture for the Ixia sessions ([dentproject#115](https://github.com/dentproject/testing/issues/115)). Can we add it in this PR?

I could add it, but I won't be able to verify the changes right now since I don't have access to the setups with Ixia.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py (hunk 9 above)
Comment (SerhiyBoikoPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: It would be nice to have, yes. If we only need it for IPv4 tests then we can add it to an ipv4 conftest file. But if other tests will also require it then we should add it to tb_utils.

Comment (yuriilisovskyi) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: I assume it may be useful for other tests as well. Is there any reference I could use to add IPv4 cleanup function?

**Comment by SerhiyBoikoPLV** [comment]
> I could add it, but I won't be able to verify the changes right now since I don't have access to the setups with Ixia.

It is a matter of adding a single function to the cleanup fixture:
```python
await tgen_utils_stop_protocols(tgen_dev)
```
But you will need to somehow get the tgen dev:
```python
tgen_dev, dent_devices = await tgen_utils_get_dent_devices_with_tgen(testbed, [], 4)
```
When you are done I can check that the fixture works as intended.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 10]
```diff
@@ -73,6 +82,73 @@ async def tb_clean_config(testbed):
...
+        parse_output=True
+    )
+    vrfs_info = out[0][dev.host_name]["parsed_output"]
+    if all(vrfs_info):
+        await IpLink.delete(input_data=[{dev.host_name: [
+            {"device": vrf_obj["ifname"]} for vrf_obj in vrfs_info
+        ]}])
+
+
+async def tb_post_test_cleanup_routes(testbed):
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: should this be `addrs` instead of `routes`?

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 11]
```diff
@@ -73,6 +82,73 @@ async def tb_clean_config(testbed):
...
+    """Removes all routes created during test. 
+    Can be used separately or by using `cleanup_routes` fixture."""
+    logger = AppLogger(DEFAULT_LOGGER)
+    dev = await tb_get_all_devices(testbed)
+    dev = dev[0]
+    logger.info("Deleting routes")
+    tgen_dev, _ = await tgen_utils_get_dent_devices_with_tgen(testbed, [], 4)
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: you can get `dev` from here

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 12]
```diff
@@ -3,10 +3,13 @@
 import os
 import time
 import re
+import re
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: re re

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/conftest.py [hunk 13]
```diff
@@ -1,14 +1,26 @@
+import asyncio
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/test/conftest.py:1]: you don't need asyncio anymore

**Review by SerhiyBoikoPLV — APPROVED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 14]
```diff
@@ -15,6 +17,12 @@
 from dent_os_testbed.lib.os.service import Service
 from dent_os_testbed.lib.os.system import System
 from dent_os_testbed.lib.tc.tc_qdisc import TcQdisc
+from dent_os_testbed.logger.Logger import AppLogger
+from dent_os_testbed.utils.test_utils.tgen_utils import (
+    tgen_utils_get_dent_devices_with_tgen,
+    tgen_utils_stop_protocols,
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: btw, this is not used anywhere

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 15]
```diff
@@ -15,6 +17,12 @@
 from dent_os_testbed.lib.os.service import Service
 from dent_os_testbed.lib.os.system import System
 from dent_os_testbed.lib.tc.tc_qdisc import TcQdisc
+from dent_os_testbed.logger.Logger import AppLogger
+from dent_os_testbed.utils.test_utils.tgen_utils import (
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: due to circular import, this should be changed to:
```python
import dent_os_testbed.utils.test_utils.tgen_utils as tgen_utils
```
And in tgen_utils.py the import should be changed to:
```python
import dent_os_testbed.utils.test_utils.tb_utils as tb_utils
```
But maybe these `post_test_cleanup` functions should be moved to a different file? For example in conftest.py besides their respective fixtures?

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py (hunk 14 above)
Comment (SerhiyBoikoPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: disregard

**Review by SerhiyBoikoPLV — CHANGES_REQUESTED** [review]
@anton7811, what do you think about moving `post_test_cleanup` function to conftest.py?

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 16]
```diff
@@ -73,6 +81,72 @@ async def tb_clean_config(testbed):
...
+        parse_output=True
+    )
+    vrfs_info = out[0][dev.host_name]["parsed_output"]
+    if all(vrfs_info):
+        await IpLink.delete(input_data=[{dev.host_name: [
+            {"device": vrf_obj["ifname"]} for vrf_obj in vrfs_info
+        ]}])
+
+
+async def tb_post_test_cleanup_addrs(testbed):
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: tb_post_test_cleanup_addrs -> tb_post_test_cleanup_ip_addrs

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 17]
```diff
@@ -73,6 +81,72 @@ async def tb_clean_config(testbed):
...
+    """Removes all IP addresses configured during test. 
+    Can be used separately or by using `cleanup_routes` fixture."""
+    logger = AppLogger(DEFAULT_LOGGER)
+    logger.info("Deleting routes")
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: You are deleting IP addresses. Connected routes will be deleted automatically. But static routes will stay as is... Good point. How to remove them?

**Review by anton7811 — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py [hunk 18]
```diff
@@ -15,6 +15,7 @@
 from dent_os_testbed.lib.os.service import Service
 from dent_os_testbed.lib.os.system import System
 from dent_os_testbed.lib.tc.tc_qdisc import TcQdisc
+
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/tb_utils.py]: please revert this change

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py [hunk 19]
```diff
@@ -0,0 +1,108 @@
...
+
+async def get_initial_routes(testbed):
+    """Gets routes defined before test. Needed to cleanup routes configured during the test"""
+    dev = await tb_get_all_devices(testbed)
+    dev = dev[0]
+    out = IpRoute.show(
+        input_data=[{dev.host_name: [{"cmd_options": "-j"}]}], 
+        parse_output=True
+    )
+    return out[0][dev]["parsed_output"]
```
Comment (SerhiyBoikoPLV) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py]: if `IpRoute.show()` fails you will not have any `parsed_output` (which is unlikely, but still)

**Review by SerhiyBoikoPLV — APPROVED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    dev = dev[0]
+    logger.info("Clearing TC")
+    out = await TcQdisc.show(
+        input_data=[{dev.host_name: [{"options": "-j"}]}], 
+        parse_output=True
+    )
+    qdiscs_info = out[0][dev.host_name]["parsed_output"]
+    for qdisc_obj in qdiscs_info:
+        if qdisc_obj["kind"] != "noqueue":
+            await TcQdisc.delete(
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py:23]: Please form a list of commands firs and then call all of them in a single TcQdisc.delete()

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+from dent_os_testbed.logger.Logger import AppLogger
+from dent_os_testbed.utils.test_utils.tb_utils import (
+    tb_get_all_devices
+)
+from dent_os_testbed.utils.test_utils.tgen_utils import (
+    tgen_utils_get_dent_devices_with_tgen
+)
+
+
+async def post_test_cleanup_qdiscs(testbed):
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py]: I think "post_test_" prefix is not necessary here. It's not forbidden to call this at any stage, even as a test step.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+    dev = await tb_get_all_devices(testbed)
+    dev = dev[0]
+    out = IpRoute.show(
+        input_data=[{dev.host_name: [{"cmd_options": "-j"}]}], 
+        parse_output=True
+    )
+    return out[0][dev]["parsed_output"]
+
+
+async def post_test_cleanup_routes(testbed, initial_routes):
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py]: Please rename "initial_routes" to "routes". So it culd be any list of routes that should be removed.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py [hunk 23]
```diff
@@ -0,0 +1,108 @@
...
+    Can be used separately or by using `cleanup_addrs` fixture."""
+    logger = AppLogger(DEFAULT_LOGGER)
+    logger.info("Deleting IP addresses from interfaces")
+    tgen_dev, dent_dev = await tgen_utils_get_dent_devices_with_tgen(testbed, [], 4)
+    dev = dent_dev[0]
+    ports = tgen_dev.links_dict[dev.host_name][1]
+    await IpAddress.flush(input_data=[{dev.host_name: [{"dev": port} for port in ports]}])
+
+
+async def get_initial_routes(testbed):
```
Comment (anton7811) [inline DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py]: I would pass dev instead of testbed here. In the setups with multiple DUTs, it will be unclear why do you process only the first one.
And in the cleanup fixture, you should do a loop against the devices list. OR you can return dictionary with device name as a key.

**Review by anton7811 — CHANGES_REQUESTED** [review]

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py (hunk 20 above)
Comment (SerhiyBoikoPLV) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py:23]: The issue with a list of commands is if one command fails then the other commands will not be executed. One such scenario is if we have a root qdisc (ets) and a bunch of leaves (tbf/red). If the root qdisc is deleted first then the leaves will be automatically deleted and the next (and subsequent) `tc qd del` command will fail.

File: DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py (hunk 22 above)
Comment (yuriilisovskyi) [reply DentOS_Framework/DentOsTestbed/src/dent_os_testbed/utils/test_utils/cleanup_utils.py]: In this case, initial_routes are the routes that we don't want to clean up (those that existed before the test started).

**Review by SerhiyBoikoPLV — APPROVED** [review]
